File: python/hackerrank/array/2d_array_ds.py
# https://www.hackerrank.com/challenges/2d-array/problem

def hourglassSum(arr):
    position_x, position_y = 0,0
    maximum_sum = -63
    wall_x = len(arr[0])
    wall_y = len(arr)

    while position_x + 2 < wall_x:
        position_y = 0
        while position_y + 2 < wall_y:
            first_row = arr[position_y][position_x:position_x + 3]
            second_row = arr[position_y + 1][position_x + 1]
            last_row = arr[position_y + 2][position_x:position_x + 3]
            current_sum = sum(first_row) + second_row + sum(last_row)
            maximum_sum = max(current_sum, maximum_sum)
            position_y += 1
        position_x += 1

    return maximum_sum

File: go/hackerrank/archive/20.08/2dArrayDs.go
// https://www.hackerrank.com/challenges/2d-array/problem

func hourglassSum(arr [][]int32) int32 {
    maximum_sum := int32(-63)

    for position_y := 0; position_y <= 3; {
        for position_x := 0; position_x <= 3; {
            current_sum := int32(0)

            selected := arr[position_y][position_x:position_x+3]
            for _, num := range(selected) {
                current_sum += num
            }

            current_sum += arr[position_y+1][position_x+1]

            selected = arr[position_y+2][position_x:position_x+3]
            for _, num := range(selected) {
                current_sum += num
            }
            if current_sum > maximum_sum {
                maximum_sum = current_sum
            }
            position_x += 1
        }
        position_y += 1
    }
    return maximum_sum
}
